### PR TITLE
Feature/#254 add qr code

### LINKED ISSUE #254: As a lecturer I want a screen where I can display a QR code so that my students can scan that code
AC:
- The QR code is displayed in the same screen where the key for the lecture is displayed
- The QR code is big enough that students can scan it when I put it on the screen
- When the key changes the QR code changes
- When the QR code is scanned the resulting text is a "valid" request as needed for #253

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
# Resolves #254 #253 

This PR adds a QR Code that automatically enrolls the user into the lecture. The user is also added to the course if he isn't already part of it.

Acceptance Criteria:
- An endpoint exists which accepts get requests
- After I send a "valid" request to that endpoint I can enter the corresponding lecture without needing to enter a key
- After I send a "valid" request to that endpoint I count as if I had entered the key manually
- Sending a "valid" request is nontrivial, I can't guess what a valid request looks like
- When I'm not logged in and access the endpoint I'm redirected to the login-page
- OPTIONAL: After logging in I'm logged into the lecture and redirected to it
- The URL I scan redirects me to the course I just logged into
- The QR code is displayed in the same screen where the key for the lecture is displayed
- The QR code is big enough that students can scan it when I put it on the screen
- When the key changes the QR code changes
- When the QR code is scanned the resulting text is a "valid" request as needed for #253

Steps needed for migration:
- need additional gems via bundler

---

Definition Of Done:
- ~~[ ] [Class-Diagram](https://www.lucidchart.com/invitations/accept/705a122b-58a9-4dfa-8b16-c936bc0a46bc) updated~~

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -135,3 +135,5 @@ gem "action-cable-testing"
 gem "icalendar"
 # Easily schedule tasks in ruby <3
 gem "rufus-scheduler"
+
+gem "rqrcode", "~> 1.1"
```
Comment (JoeAtHPI) [inline Gemfile]: keep consistency with the rest of the files by adding a comment above the gem and what it does

File: app/controllers/lectures_controller.rb [hunk 2]
```diff
@@ -176,4 +185,26 @@ def validate_course_creator
         redirect_to @course, alert: "You can only access your own courses."
       end
     end
+
+    def generate_enrollment_qr_code
+      enrollment_url = request.base_url + "/courses/" + @course.id.to_s + "/lectures/" + @lecture.id.to_s + "/enroll"
```
Comment (JoeAtHPI) [inline app/controllers/lectures_controller.rb]: uuuh, hardcoded stuff inside a controller. i'm sure there must be a rails routes for what you're accessing here

File: config/routes.rb [hunk 3]
```diff
@@ -2,6 +2,7 @@
   get "/courses/:course_id/lectures/current", to: "lectures#current", as: "current_lectures"
   get "/ical/:hash_id", to: "ical#show", as: "ical"
   post "/courses/:course_id/lectures/join_lecture", to: "lectures#join_lecture", as: "join_lecture"
+  get "/courses/:course_id/lectures/:lecture_id/enroll", to: "lectures#join_lecture_with_url", as: "join_lecture_with_url"
```
Comment (JoeAtHPI) [inline config/routes.rb]: there it is ;)

File: spec/controllers/lectures_controller_spec.rb [hunk 4]
```diff
@@ -78,7 +82,9 @@
 
     it "redirects to course overview for not joined students", :logged_lecturer do
       lecture = Lecture.create! valid_attributes_with_lecturer_with_course
-      login_student()
+      user = FactoryBot.create(:user, :student)
```
Comment (JoeAtHPI) [inline spec/controllers/lectures_controller_spec.rb]: sure you cant extract the user you're creating and logging in as them? would result in less duplicated lines

File: spec/controllers/lectures_controller_spec.rb [hunk 5]
```diff
@@ -370,7 +376,7 @@
         sign_out(@user)
         @user1 = FactoryBot.create(:user, :student)
         login_student(@user1)
-        @lecture.join_lecture(@user1)
+        join_course_and_lecture(@user1, @lecture)
```
Comment (JoeAtHPI) [inline spec/controllers/lectures_controller_spec.rb]: same here

File: spec/views/lectures/show.html.erb_spec.rb [hunk 6]
```diff
@@ -43,6 +43,13 @@
...
+    it "renders enrollment qr code if enrollment key is present" do
+      @qr_code = RQRCode::QRCode.new("http://some-random.url/that/is/not/tested")
+      render
+      assert_select "a", "Enrollment Key"
+      expect(rendered).to have_selector("div", :class => "qr-code-container")
+    end
+
     it "does not render enrollment key tab button if enrollment key is not present" do
       @lecture.update(enrollment_key: nil)
       render
```
Comment (JoeAtHPI) [inline spec/views/lectures/show.html.erb_spec.rb]: no bdd test for scanning in a qr code and then checking if the user is redirected and saved as a participant to that corresponding lecture?

Comment (MichaelBuessemeyer) [reply spec/views/lectures/show.html.erb_spec.rb]: Sorry but at test for scanning a qrcode is in my opinion too hard and out of scope

**Comment by MichaelBuessemeyer** [comment]
@grittaweisheit @JoeAtHPI I tried adding a test, that checks whether the `join_lecture_with_url` works. I cannot get the test working. Can you fix this? To see the specific ttest, just look at the changes of commit https://github.com/hpi-swt2/lecture-portal/pull/288/commits/702fd24cfeedfba16f7464fefe567cae4823afef

File: spec/controllers/lectures_controller_spec.rb (hunk 4 above)
Comment (MichaelBuessemeyer) [reply spec/controllers/lectures_controller_spec.rb]: No this is not possible, the `login_student` function creates a student if no student is given but does not return is

**Pull request merged**